lang/cs/Org.Apache.REEF.Bridge.Core.Grpc/Driver/DriverClientService.cs

```diff
@@ -0,0 +1,734 @@
+            IInjector injector,
+            BridgeClock bridgeClock,
+            DriverServiceClient driverServiceClient)
+        {
+            _receivedStartEvent = false;
+            _injector = injector;
+            _bridgeClock = bridgeClock;
+            _driverServiceClient = driverServiceClient;
+            _server = new Server
+            {
+                Services = { DriverClient.BindService(this) },
+                Ports = { new ServerPort("localhost", 0, ServerCredentials.Insecure) }
+            };
+            _server.Start();
+            foreach (var serverPort in _server.Ports)
+            {
+                Logger.Log(Level.Info, "Server port {0}", serverPort.BoundPort);
+                _serverPort = serverPort.BoundPort;
+            }
+
+            Logger.Log(Level.Info, $"Client service started on port {_serverPort}");
+        }
+
+        #region Driver Client Service implementation
+
+        public void Start()
+        {
+            try
+            {
+                // could throw application level exception
+                _driverBridge = _injector.GetInstance<DriverBridge>();
+                _driverServiceClient.RegisterDriverClientService("localhost", _serverPort);
+            }
+            catch (Exception ex)
+            {
+                Logger.Log(Level.Error, "Driver application creation error", ex);
+                _driverServiceClient.RegisterDriverClientService(ex);
+                _server.ShutdownAsync();
+            }
+        }
+
+        public void AwaitTermination()
+        {
+            Logger.Log(Level.Info, "Awaiting client server termination");
+            _server.ShutdownTask.Wait();
+            Logger.Log(Level.Info, "Client server terminated");
+        }
+
+        #endregion
+
+        public override async Task<Void> AlarmTrigger(AlarmTriggerInfo request, ServerCallContext context)
+        {
+            await _bridgeClock.OnNextAsync(new BridgeClock.AlarmInfo
+            {
+                AlarmId = request.AlarmId,
+                Timestamp = request.Timestamp
+            });
+            return Void;
+        }
+
+        public override Task<IdleStatus> IdlenessCheckHandler(Void request, ServerCallContext context)
+        {
+            return Task.FromResult(new IdleStatus
+            {
+                IsIdle = IsIdle,
+                Reason = "Driver client"
+            });
+        }
+
+        public override async Task<Void> StartHandler(StartTimeInfo request, ServerCallContext context)
+        {
+            try
+            {
+                Logger.Log(Level.Info, "Received start event at time {0}", request.StartTime);
+                await _driverBridge.DispatchStartEventAsync(new BridgeDriverStarted(new DateTime(request.StartTime)));
+                _receivedStartEvent = true;
+            }
+            catch (Exception ex)
+            {
+                _bridgeClock.Dispose(ex);
+            }
+            return Void;
+        }
```

Comment:
Looks like we are using the same `try`/`catch`/`return Void;` block in many methods. Maybe we can define a function like
```C#
private Task<Void> StopClockOnError(string where, Action action)
{
    try
    {
        action();
    }
    catch (Exception ex)
    {
        Logger.Log(Level.Warning, "Stopping event loop. Exception at " + where, ex);
        _bridgeClock.Dispose(ex);
    }

    return Task.FromResult(Void);
}
```
so we can write more compact e.g.
```C#
public override async Task<Void> StartHandler(StartTimeInfo request, ServerCallContext context)
{
    return await StopClockOnError("Start " + request.StartTime, async () =>
    {
        Logger.Log(Level.Info, "Received start event at time {0}", request.StartTime);
        await _driverBridge.DispatchStartEventAsync(
            new BridgeDriverStarted(new DateTime(request.StartTime)));
        _receivedStartEvent = true;
    });
}
```
(not 100% sure if it is worth it, though :)

Reply:
I'm not sure this is worth doing either :)